Mathlib/Combinatorics/Additive/FreimanHom.lean

```diff
@@ -289,6 +291,50 @@ lemma IsMulFreimanHom.inv (hf : IsMulFreimanHom n A B f) : IsMulFreimanHom n A B
 
 end DivisionCommMonoid
 
+namespace Fin
+variable {k m n : ℕ}
+
+private lemma aux (hm : m ≠ 0) (hkmn : m * k ≤ n) : k < (n + 1) :=
+  Nat.lt_succ_iff.2 $ le_trans (Nat.le_mul_of_pos_left _ hm.bot_lt) hkmn
+
+/-- **No wrap-around principle**.
+
+The first `k + 1` elements of `Fin (n + 1)` are `m`-Freiman isomorphic to the first `k + 1` elements
+of `ℕ` assuming there is no wrap-around. -/
+lemma isAddFreimanIso_Iic (hm : m ≠ 0) (hkmn : m * k ≤ n) :
+    IsAddFreimanIso m (Iic (k : Fin (n + 1))) (Iic k) val where
+  bijOn.left := by simp [MapsTo, Fin.le_iff_val_le_val, Nat.mod_eq_of_lt, aux hm hkmn]
+  bijOn.right.left := val_injective.injOn _
+  bijOn.right.right x (hx : x ≤ _) :=
+    ⟨x, by simpa [le_iff_val_le_val, -val_fin_le, Nat.mod_eq_of_lt, aux hm hkmn, hx.trans_lt]⟩
+  map_sum_eq_map_sum s t hsA htA hs ht := by
+    have (u : Multiset (Fin (n + 1))) : Nat.castRingHom _ (u.map val).sum = u.sum := by simp
+    rw [← this, ← this]
+    have {u : Multiset (Fin (n + 1))} (huk : ∀ x ∈ u, x ≤ k) (hu : card u = m) :
+        (u.map val).sum < (n + 1) := Nat.lt_succ_iff.2 $ hkmn.trans' $ by
+      rw [← hu, ← card_map]
+      refine sum_le_card_nsmul (u.map val) k ?_
+      simpa [le_iff_val_le_val, -val_fin_le, Nat.mod_eq_of_lt, aux hm hkmn] using huk
+    exact ⟨congr_arg _, CharP.natCast_injOn_Iio _ (n + 1) (this hsA hs) (this htA ht)⟩
+
+/-- **No wrap-around principle**.
+
+The first `k` elements of `Fin (n + 1)` are `m`-Freiman isomorphic to the first `k` elements of `ℕ`
+assuming there is no wrap-around. -/
+lemma isAddFreimanIso_Iio (hm : m ≠ 0) (hkmn : m * k ≤ n) :
+    IsAddFreimanIso m (Iio (k : Fin (n + 1))) (Iio k) val := by
+  obtain _ | k := k
+  · simp [← bot_eq_zero]; simp [← _root_.bot_eq_zero, -bot_eq_zero']
```

Comment:
This looks kinda brittle. Would a simp only or simp_rw be clearer?

Reply:
No, it's really not brittle. The issue is that simp in Lean 4 really dumbly does not remove "forward `foo`" when you ask it to use "backward `foo`": If `foo` is a simp lemma then `simp [← foo]` will invariably loop (unless you're lucky and it closes the goal before hitting the loop.

Reply:
https://github.com/leanprover/lean4/issues/4290